### PR TITLE
fix: Rounding issue in approval amount (use site suggestion)

### LINKED ISSUE #8199: [Bug]: Rounding issue in approval amount (use site suggestion)
### Describe the bug

While testing our dApp using metamask webview / metamask android, I saw that the approval transaction amount is modified by metamask even if the user clicks on "Use site suggestion".
I believe there is a rounding issue when metamask decodes and encodes the user amount.

For example, this is the tx data which is generated by dApp:
```
0x095ea7b300000000000000000000000069460570c93f9de5e2edbc3052bf10125f0ca22d000000000000000000000000000000000000000000000000576129d2d21d64a5
```

and this is the final data used by metamask even if user clicks on use site suggestion:

```
0x095ea7b300000000000000000000000069460570c93f9de5e2edbc3052bf10125f0ca22d000000000000000000000000000000000000000000000000576129d2d21d6400
```

as you can see, the amount part is slightly changed. 
Expected amount based on original data: 6.2963597394856**76709**
Metamask modified version: 6.2963597394856**76544**

### Screenshots/Recordings

None

### Expected behavior

I think when user clicks on use site suggestion, dapp transaction data should not be changed.

### Steps to reproduce

1. Go to Metamask Webview
2. Send an approval transaction with this 'data' (0x095ea7b300000000000000000000000069460570c93f9de5e2edbc3052bf10125f0ca22d000000000000000000000000000000000000000000000000576129d2d21d6**4a5**)
3. Click on use site suggested amount and accept the approval transaction.
4. Final transaction data is changed by mistake to this  one:
(0x095ea7b300000000000000000000000069460570c93f9de5e2edbc3052bf10125f0ca22d000000000000000000000000000000000000000000000000576129d2d21d6**400**)

### Error messages or log output

There is no specific error log. 

### Version

7.12.3

### Build type

None

### Device

Samsung S21 

### Operating system

Android

### Additional context

I faced this bug in Metamask android version several times, but I'm not sure if it is only related to Metamask mobile or the android version. 

### Severity

It creates some bugs during the next steps of dapps transactions flow.

### PR DESCRIPTION
**Pull request by blackdevelopa**
## **Description**

There is an [issue](https://github.com/MetaMask/metamask-mobile/issues/8199) that shows that really large dapp proposed value is rounded off for approve transaction. Noticed that this happens when hexToBN value is wrapped using the Number function. This PR specifies add a boolean to prevent this  for approve transactions while other use cases should act as it is.

## **Related issues**

Fixes: #8199 

## **Manual testing steps**

1. Enable input customisation for erc20 token. [See](https://github.com/MetaMask/metamask-mobile/issues/8199#issuecomment-1886096624)
2. Enter a really large value
3. See it doesn't round off the dapp proposed value

### **Before**
https://recordit.co/l0DWAilLla

### **After**
https://recordit.co/gEKmKf9Yq7

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/util/number/index.test.ts [hunk 1]
```diff
@@ -130,6 +130,27 @@ describe('Number utils :: fromTokenMinimalUnit', () => {
     expect(fromTokenMinimalUnit(1e16, 6)).toEqual('10000000000');
     expect(fromTokenMinimalUnit(1e18, 18)).toEqual('1');
   });
+
+  it('fromTokenMinimalUnit using BN number and not round number', () => {
```
Comment (segun) [inline app/util/number/index.test.ts:134]: can we split this test into 2, 

1. Tests that value is rounded when `rounded` is true
2.Tests that value is not rounded when `rounded` is false

reference: [keep-tests-focused ](https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md#keep-tests-focused )

**Review by jpuri — DISMISSED** [review]
Will be n ice to have more testing on this change.

File: app/util/number/index.test.ts [hunk 2]
```diff
@@ -130,6 +130,31 @@ describe('Number utils :: fromTokenMinimalUnit', () => {
...
+      '10000',
+    );
+    expect(
+      fromTokenMinimalUnit(new BN('50000000000000000000000'), 18, false),
+    ).toEqual('50000');
+
+    expect(
+      fromTokenMinimalUnit(hexToBN('576129d2d21d64a5'), 18, false),
+    ).toEqual('6.296359739485676709');
+  });
```
Comment (digiwand) [inline app/util/number/index.test.ts:157]: [suggestion] I think we can further enhance this by testing the same value inputs with the rounding. 
follow-up to https://github.com/MetaMask/metamask-mobile/pull/8485#discussion_r1476102686.

e.g.
```
  it('should round number by default', () => {
    expect(fromTokenMinimalUnit(new BN('1000000000000000000'), 18)).toEqual(
      '1',
    );
    expect(fromTokenMinimalUnit(new BN('10000000000000000000'), 18)).toEqual(
      '10',
    );
    expect(fromTokenMinimalUnit(new BN('100000000000000000000'), 18)).toEqual(
      '100',
    );
    expect(fromTokenMinimalUnit(new BN('1000000000000000000000'), 18)).toEqual(
      '1000',
    );
    expect(fromTokenMinimalUnit(new BN('10000000000000000000000'), 18)).toEqual(
      '10000',
    );

    // test decimal greater than 30,000
    expect(
      fromTokenMinimalUnit(new BN('50000000000000000000000'), 18),
    ).toEqual('49999.999999999995805696');
    
    // test decimal less than 1e-14
    expect(
      fromTokenMinimalUnit(hexToBN('576129d2d21d64a5'), 18),
    ).toEqual('6.296359739485676544');
  });

  it('should not round number if rounded is false', () => {
    expect(fromTokenMinimalUnit(new BN('1000000000000000000'), 18, false)).toEqual(
      '1',
    );
    expect(fromTokenMinimalUnit(new BN('10000000000000000000'), 18, false)).toEqual(
      '10',
    );
    expect(fromTokenMinimalUnit(new BN('100000000000000000000'), 18, false)).toEqual(
      '100',
    );
    expect(fromTokenMinimalUnit(new BN('1000000000000000000000'), 18, false)).toEqual(
      '1000',
    );
    expect(fromTokenMinimalUnit(new BN('10000000000000000000000'), 18, false)).toEqual(
      '10000',
    );

    // test decimal greater than 30,000
    expect(
      fromTokenMinimalUnit(new BN('50000000000000000000000'), 18, false),
    ).toEqual('50000');

    // test decimal less than 1e-14
    expect(
      fromTokenMinimalUnit(hexToBN('576129d2d21d64a5'), 18, false),
    ).toEqual('6.296359739485676709');
  });
```

File: app/util/number/index.js [hunk 3]
```diff
@@ -93,14 +93,16 @@ export function fromWei(value = 0, unit = 'ether') {
...
+ * Included a rounded parameter to avoid rounding the number for really large numbers and backwards compatibility
  *
  * @param {number|string|Object} minimalInput - Token minimal unit to convert
  * @param {number|string} decimals - Token decimals to convert
  * @returns {string} - String containing the new number
  */
-export function fromTokenMinimalUnit(minimalInput, decimals) {
-  minimalInput = addHexPrefix(Number(minimalInput).toString(16));
-  let minimal = safeNumberToBN(minimalInput);
+export function fromTokenMinimalUnit(minimalInput, decimals, rounded = true) {
```
Comment (digiwand) [inline app/util/number/index.js:102]: Q: what purpose do `rounded = true` and `Number(minimalInput)` have here?

**Review by jpuri — DISMISSED** [review]

File: app/util/number/index.test.ts (hunk 2 above)
Comment (blackdevelopa) [reply app/util/number/index.test.ts:188]: Thank you. Updated.

File: app/util/number/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (blackdevelopa) [reply app/util/number/index.js]: Thank you. Updated

File: app/util/number/index.test.ts [hunk 4]
```diff
@@ -130,6 +130,62 @@ describe('Number utils :: fromTokenMinimalUnit', () => {
     expect(fromTokenMinimalUnit(1e16, 6)).toEqual('10000000000');
     expect(fromTokenMinimalUnit(1e18, 18)).toEqual('1');
   });
+
+  it('should round number by default', () => {
```
Comment (segun) [inline app/util/number/index.test.ts:134]:
```suggestion
  it('rounds number by default', () => {
```
reference: https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md#use-it-to-specify-the-desired-behavior-for-the-code-under-test

File: app/util/number/index.test.ts [hunk 5]
```diff
@@ -130,6 +130,62 @@ describe('Number utils :: fromTokenMinimalUnit', () => {
...
+      '49999.999999999995805696',
+    );
+
+    // test decimal less than 1e-14
+    expect(fromTokenMinimalUnit(hexToBN('576129d2d21d64a5'), 18)).toEqual(
+      '6.296359739485676544',
+    );
+  });
+
+  it('should not round number if isRounding is false', () => {
```
Comment (segun) [inline app/util/number/index.test.ts:162]:
```suggestion
  it('does not round number if isRounding is false', () => {
```
Reference: https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md#use-it-to-specify-the-desired-behavior-for-the-code-under-test

**Review by segun — COMMENTED** [review]
Few comments to follow guidelines.

**Review by segun — APPROVED** [review]

**Review by jpcloureiro — APPROVED** [review]

**Pull request merged**